CONTRIBUTING.md

```diff
@@ -0,0 +1,63 @@
+
+Inside the Github Action referred to by `uses`, there must be an `action.yml` file. Let's check out the `verify-pull-request` action file here: https://github.com/Workday/canvas-kit-actions/blob/main/verify-pull-request/action.yml
+
+```yaml
+name: 'Verify Pull Request'
+description: 'Verify the pull request title and metadata to ensure correct changeset results'
+author: 'Workday'
+inputs:
+  token:
+    required: true
+    description: 'A github token with at least read access to the repository'
+  pullRequestNumber:
+    required: false
+    description: 'Optional manual override of a pull request number. Useful for workflow dispatch jobs.'
+runs:
+  using: 'node12'
+  main: '../dist/actions/verifyPullRequest.js'
+```
+
+The `runs` block contains the `main` block which points to the source file. To find this file, transpose the `dist` with `src`. For example, here's the source code for this action: https://github.com/Workday/canvas-kit-actions/blob/main/src/actions/verifyPullRequest.ts.
+
+## Navigating code
+
+All actions import from `repo` and `utils`:
+
+- `repo`: Contains all side-effect API calls to the GitHub API or the file system. All API responses are strongly typed using GitHub's rest API or GraphQL and using https://github.com/Quramy/ts-graphql-plugin to generate Typescript types from a GraphQL query. There are no tests for these functions as they only return API calls. Some of the more complex API responses are stored as fixture data in the `fixtures` directory. These fixture JSON files help to understand the returned API calls as well as are used in some tests. If you make changes to this file, run `yarn ts-graphql-plugin typegen` to regenerate Typescript types from the GraphQL schema.
+- `utils`: Contains only pure functions to manipulate data or responses from API calls. All functions here are unit tested. Most fixes and code changes can be made here with 100% confidence that the CI will still work. The unit tests also help to understand data flow.
+
+The combination of strong types and unit testing ensures a high degree of confidence in changes without having to test changes via CI runs triggered by real GitHub events in the http://github.com/Workday/canvas-kit repository.
+
+## Contributing
+
+The best thing to do is to create a test in `src/utils.test.ts` to verify the change you'd like
+to make. Create a function in `src/utils.ts` that manipulates the data the way you want. If you
+need to also create or modify API calls, only do this from the `src/repo.ts` file. You can use the [GitHub GraphQL explorer](https://docs.github.com/en/graphql/overview/explorer) to make GraphQL requests to get the data you want.
+
+If you want to create a new action, create a new folder in the root AND in the `src/actions`.
+For example, if we wanted to create a new action called `verify-issue`, create the following files:
+
+- `verify-issue/action.yml` - entry point
+- `src/actions/verifyIssue.ts` - The `action.yml` file points here
```

Comment:
Do you have a way to run the test?

Reply:
I don't have a way to test side effects locally. I could create a core wrapper that could be used locally that creates a fixture for Github events, but everything in the `actions` file is strongly typed and will have side effects that would make changes to Github.

This is why `utils` is separated from `actions` and `repo`. Side effects and pure functions are separated to isolate testing. The way I test side effects is on my fork of Canvas Kit where I create a workflow in my fork that references a new action and test there. Without a lot of extra wrapping, that's as good as we can get for testing outside of production Canvas Kit.